### PR TITLE
Add binning with CONCOCT

### LINKED ISSUE #339: Add CONCOCT as an alternative binner
### Description of feature

https://github.com/BinPro/CONCOCT

At the request of @alexhbnr

### PR DESCRIPTION
**Pull request by jfy133**
Closes #339 

Note CONCOCT seems to be quite permissive (doesn't itself produce any 'failed' bins, so produces _a lot_ of bins, even after SPLIT_FASTA). I'm not yet sure if this will cause an issue with our CI tests, but lets see.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/modules.config
Comment (jfy133) [inline conf/modules.config:438]:
```suggestion
    }
```

File: subworkflows/local/binning.nf
Comment (jfy133) [inline subworkflows/local/binning.nf:120]:
```suggestion
        FASTA_BINNING_CONCOCT ( ch_concoct_input )
```

File: docs/output.md
Comment (d4straub) [inline docs/output.md:360]: > If DAS Tool does not find any bins passing your selected threshold it will exit with an error files.

Wondering if that is a typo or will there be indeed an error file?

File: subworkflows/local/binning.nf
Comment (d4straub) [inline subworkflows/local/binning.nf:21]:
```suggestion
```

File: subworkflows/local/binning_refinement.nf
Comment (d4straub) [inline subworkflows/local/binning_refinement.nf:45]:
```suggestion
                                            concoct:  it[0]['binner'] == 'CONCOCT'
```

File: subworkflows/local/binning_refinement.nf
Comment (d4straub) [inline subworkflows/local/binning_refinement.nf:70]: Is that still a problem?

File: workflows/mag.nf
Comment (d4straub) [inline workflows/mag.nf:564]: Do I understand this line correct that refinement with dastool will only be used when all three binners were not skipped? While I agree that it is a good idea to use all 3 binners for refinement, shouldnt it be possible to also use refinement with only 2 binners?

File: docs/output.md
Comment (jfy133) [inline docs/output.md:360]:
```suggestion
> ⚠️ If DAS Tool does not find any bins passing your selected threshold it will exit with an error. Such an error is 'ignored' by nf-core/mag, therefore you will not find files in the `GenomeBinning/DASTool/` results directory for that particular sample.
```

Oops left over from a re-phrasing

File: subworkflows/local/binning_refinement.nf
Comment (jfy133) [reply subworkflows/local/binning_refinement.nf:70]: It wasn't an issue with our pipeline but a bug in DAS_Tool and in this case only effects CONCOCT:

https://github.com/cmks/DAS_Tool/issues/90

I'm waiting to see if there will be a quick patch release for it in the next day or so (I wanted the rest of this PR reviewed in the meantime), if not I'll go back and fix the header names coming out of CONCOCT so they don't have spaces in them.

I just noticed I forgot to put this to draft... will do that now

Comment (jfy133) [reply subworkflows/local/binning_refinement.nf:70]: While there is some activity on DAS_Tool I don't know how fast it will be. Instead I will 'fix' the MEGAHIT contig names instead

Comment (jfy133) [reply subworkflows/local/binning_refinement.nf]: I've added a 'fix' for the MEGAHIT contig names (which ended up being more complicated due to PROKKA contig length name restrictions...)

File: workflows/mag.nf
Comment (jfy133) [reply workflows/mag.nf]: Done

File: workflows/mag.nf
Comment (d4straub) [inline workflows/mag.nf:565]:
```suggestion
        if ( params.refine_bins_dastool && [ params.skip_metabat2, params.skip_maxbin2, params.skip_concoct ].count(false) > 1 ) {
```
I believe this would be better, not an exponential increase of tests, but linear. I tested it with nextflow console, but with variables instead of params, but I expect it would work?

**Review by d4straub — APPROVED** [review]
Looks good to me.

File: workflows/mag.nf
Comment (jfy133) [reply workflows/mag.nf:565]: Ooooh that's clever, I admittedly get stuck on how to do this! I will try to test locally before committing

**Comment by jfy133** [comment]
@d4straub Thanks for the review! I checked and your method is indeed better! I noticed this check (and another I added earlier) were in the wrong place, I've moved that to the correct place now. Could you just check for this one more time (and if the error message is good), then I'll merge once tests pass

File: lib/WorkflowMag.groovy
Comment (d4straub) [inline lib/WorkflowMag.groovy:95]: Doesnt `.count(false) <= 1` mean that 1 or less of the binners are run?
I had in mind to check whether 2 or more binners (> 1) are not skipped (false), following line 93 `Check if at least two binners were applied in order to run DAS Tool for bin refinment`:
```suggestion
        if ( params.refine_bins_dastool && [ params.skip_metabat2, params.skip_maxbin2, params.skip_concoct ].count(false) > 1 ) {
```

Comment (jfy133) [reply lib/WorkflowMag.groovy]: Ah, yeah sorry. I flipped the use case here, as I want to throw an error if it is less than 1.

But using > 1 means you don't have to adjust in the future (I just need to take the opposite to throw the error), like you suggest :+1: 

Added a comment for myself as I'm very bad at thinking about these things abstractly.

Comment (jfy133) [reply lib/WorkflowMag.groovy]: I've tested now @d4straub! If you're OK with it can you merge.

Comment (d4straub) [reply lib/WorkflowMag.groovy]: Haha true, I had it wrong here! I am happy with the change!

Comment (jfy133) [reply lib/WorkflowMag.groovy]: Ok, once tests pass I merge (ignoring the template lint stuff)

**Pull request merged**
